### PR TITLE
[WIP] RecSys Loss

### PR DESCRIPTION
**Pull request by zkid18**
The very first draft of Loss functions  for RecSys

## Before submitting (checklist)

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [ ] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [ ] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [ ] Did you check the docs with `make check-docs`?
- [ ] Did you write any new necessary tests?
- [ ] Did you check that your code passes the unit tests `pytest .` ? 
- [ ] Did you add your new functionality to the docs?
- [ ] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)?
- [ ] Did you run [colab minimal CI/CD](https://colab.research.google.com/drive/1JCGTVvWlrIsLXMPRRRSWiAstSLic4nbA) with `latest` and `minimal` requirements?

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### FAQ
Please review the FAQ before submitting an issue:
- [ ] I have read the [documentation and FAQ](https://catalyst-team.github.io/catalyst/)
- [ ] I have reviewed the [minimal examples section](https://github.com/catalyst-team/catalyst#minimal-examples)
- [ ] I have checked the [changelog](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md) for main framework updates
- [ ] I have read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)
- [ ] I have joined [Catalyst slack (#__questions channel)](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for issue discussion

### CODE REVIEW

File: catalyst/contrib/nn/criterion/recsys.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    For that pair and compare it with the ground truth. The goal for the ranker is to 
+    minimize the number of inversions in ranker. 
+
+    Input space: pairs of documents (d1, d2)
+    Output space: preferences (yes/no) for a given doc.pair
+    """
+    def __init__(self) -> None:
+        super().__init__()
+    
+    def forward(self):
```
Comment (Casyfill) [inline catalyst/contrib/nn/criterion/recsys.py:34]: shall it be `def forward(self, positive_score, negative_score)` here, just to state the intent?

File: catalyst/contrib/nn/criterion/recsys.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    
+    def forward(self, positive_score, negative_score):
+        """
+        Args
+            positive_predictions: torch.LongTensor
+                Tensor containing predictions for known positive items.
+            negative_predictions: torch.LongTensor
+                Tensor containing predictions for sampled negative items.
+        """
+        loss = -torch.log(self.gamma + torch.sigmoid(positive_score - negative_score))
```
Comment (Casyfill) [inline catalyst/contrib/nn/criterion/recsys.py:71]: `- F.logsigmoid(positive_score - negative_score) `?

File: catalyst/contrib/nn/criterion/recsys.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        highest_negative_score, _ = torch.max(negative_score, 0).squezee()
+
+        loss = torch.clamp(highest_negative_score -
+                       positive_score +
+                       1.0, 0.0)
+
+        return loss.mean()
+
+
+class LogisticLoss(Pointwise):
```
Comment (Casyfill) [inline catalyst/contrib/nn/criterion/recsys.py:107]: Pairwise, right ?

**Comment by Scitator** [comment]
merged :)
